execution/shuffle.go

```diff
@@ -0,0 +1,449 @@
+			}
+
+			switch el := element.Type.(type) {
+			case *QueueElement_Record:
+				// We got a record, so we close all the received subscriptions from the previous streams.
+				for _, sourceIndexToClose := range sourceOrder[:orderIndex] {
+					if changeSubscriptions[sourceIndexToClose] == nil {
+						continue
+					}
+					err := changeSubscriptions[sourceIndexToClose].Close()
+					if err != nil {
+						return nil, errors.Wrapf(err, "couldn't close changes subscription for source partition with index %d", sourceIndexToClose)
+					}
+				}
+
+				return el.Record, nil
+
+			case *QueueElement_Watermark:
+				if err := watermarksMap.Set(&sourcePartitionValue, el.Watermark); err != nil {
+					return nil, errors.Wrapf(err, "couldn't set new watermark value for source partition with index %d", sourcePartition)
+				}
+				continue queuePoppingLoop
+
+			case *QueueElement_EndOfStream:
+				phantom := octosql.MakePhantom()
+				if err := endOfStreamsMap.Set(&sourcePartitionValue, &phantom); err != nil {
+					return nil, errors.Wrapf(err, "couldn't set end of stream for source partition with index %d", sourcePartition)
+				}
+				continue sourcePartitionsLoop
+
+			case *QueueElement_Error:
+				return nil, errors.Wrapf(errors.New(el.Error), "received error from source partition with index %d", sourcePartition)
+
+			default:
+				panic("unreachable")
+			}
+		}
+	}
+
+	changeSubscriptionsNonNil := make([]*storage.Subscription, 0)
+	for i := range changeSubscriptions {
+		if changeSubscriptions[i] != nil {
+			changeSubscriptionsNonNil = append(changeSubscriptionsNonNil, changeSubscriptions[i])
+		}
+	}
+
+	if len(changeSubscriptionsNonNil) == 0 {
+		return nil, ErrEndOfStream
+	}
+
+	return nil, NewErrWaitForChanges(storage.ConcatSubscriptions(ctx, changeSubscriptionsNonNil...))
+}
+
+func (rs *ShuffleReceiver) GetWatermark(ctx context.Context, tx storage.StateTransaction) (time.Time, error) {
+	streamPrefixedTx := tx.WithPrefix(rs.streamID.AsPrefix())
+	watermarksMap := storage.NewMap(streamPrefixedTx.WithPrefix(watermarksPrefix))
+
+	var earliestWatermark time.Time
+
+	for sourcePartition := 0; sourcePartition < rs.sourcePartitionCount; sourcePartition++ {
+		sourcePartitionValue := octosql.MakeInt(sourcePartition)
+
+		var protoTimestamp timestamp.Timestamp
+		err := watermarksMap.Get(&sourcePartitionValue, &protoTimestamp)
+		if err == storage.ErrNotFound {
+			return time.Time{}, nil
+		} else if err != nil {
+			return time.Time{}, errors.Wrapf(err, "couldn't get watermark value for source partition %d", sourcePartition)
+		}
+
+		curWatermark, err := ptypes.Timestamp(&protoTimestamp)
+		if err != nil {
+			return time.Time{}, errors.Wrapf(err, "couldn't parse proto timestamp as time for source partition %d", sourcePartition)
+		}
+
+		if sourcePartition == 0 { // earliestWatermark is not set yet
+			earliestWatermark = curWatermark
+		} else if curWatermark.Before(earliestWatermark) {
```

Comment:
You can do multiple if conditions to prevent rewriting if body like `if sourcePArtition == 0 || curWatermark.Before(earliestWatermark) {...}`

Reply:
Imo it won't decrease readability

Reply:
The thing is, the comment describe the first if case, so that's why I wanted it this way.

The compiler should optimize this anyways.